model/trigger/build.go

```diff
@@ -156,26 +162,79 @@ func (t *buildTriggers) buildOutcome(sub *event.Subscription) (*notification.Not
 		return nil, nil
 	}
 
-	return t.generate(sub)
+	return t.generate(sub, "")
 }
 
 func (t *buildTriggers) buildFailure(sub *event.Subscription) (*notification.Notification, error) {
 	if t.data.Status != evergreen.BuildFailed {
 		return nil, nil
 	}
 
-	return t.generate(sub)
+	return t.generate(sub, "")
 }
 
 func (t *buildTriggers) buildSuccess(sub *event.Subscription) (*notification.Notification, error) {
 	if t.data.Status != evergreen.BuildSucceeded {
 		return nil, nil
 	}
 
-	return t.generate(sub)
+	return t.generate(sub, "")
 }
 
-func (t *buildTriggers) makeData(sub *event.Subscription) (*commonTemplateData, error) {
+func (t *buildTriggers) buildExceedsDuration(sub *event.Subscription) (*notification.Notification, error) {
+	if t.data.Status != evergreen.BuildSucceeded && t.data.Status != evergreen.BuildFailed {
+		return nil, nil
+	}
+	thresholdString, ok := sub.TriggerData[event.BuildDurationKey]
+	if !ok {
+		return nil, fmt.Errorf("subscription %s has no build time threshold", sub.ID)
+	}
+	threshold, err := strconv.Atoi(thresholdString)
+	if err != nil {
+		return nil, fmt.Errorf("subscription %s has an invalid time threshold", sub.ID)
+	}
+
+	maxDuration := time.Duration(threshold) * time.Second
+	if t.build.TimeTaken < maxDuration {
+		return nil, nil
+	}
+	return t.generate(sub, fmt.Sprintf("exceeded %d seconds", threshold))
+}
+
+func (t *buildTriggers) buildRuntimeChange(sub *event.Subscription) (*notification.Notification, error) {
+	if t.data.Status != evergreen.BuildSucceeded && t.data.Status != evergreen.BuildFailed {
+		return nil, nil
+	}
+	percentString, ok := sub.TriggerData[event.BuildPercentChangeKey]
+	if !ok {
+		return nil, fmt.Errorf("subscription %s has no percentage increase", sub.ID)
+	}
+	percent, err := strconv.ParseFloat(percentString, 64)
+	if err != nil {
+		return nil, fmt.Errorf("subscription %s has an invalid percentage", sub.ID)
+	}
+
+	lastGreen, err := t.build.PreviousSuccessful()
+	if err != nil {
+		return nil, errors.Wrap(err, "error retrieving last green build")
+	}
+	if lastGreen == nil {
+		return nil, nil
+	}
+	thisBuildDuration := float64(t.build.TimeTaken)
+	prevBuildDuration := float64(lastGreen.TimeTaken)
+	ratio := thisBuildDuration / prevBuildDuration
+	if !util.IsFiniteNumericFloat(ratio) {
+		return nil, nil
+	}
+	percentChange := math.Abs(100*ratio - 100)
```

Comment:
can we extract this logic out, and share it between all the time based triggers?

Reply:
If you're talking about the math parts, I think so but I'd like to do that in a separate ticket since it will touch all 3 triggers

Reply:
Yep. Sounds good to me, please ticket before merging this in.